File: hcx-pipeline-jobs/coverage-eligibility-check-job/src/test/scala/org/swasth/dp/coverageeligibilitycheck/spec/CoverageEligibilityProcessFunctionTest.scala
package org.swasth.dp.coverageeligibilitycheck.spec

import com.fasterxml.jackson.databind.ObjectMapper
import com.google.gson.Gson
import com.typesafe.config.{Config, ConfigFactory}
import org.apache.flink.api.scala.createTypeInformation
import org.apache.flink.streaming.runtime.streamrecord.StreamRecord
import org.apache.flink.streaming.util.ProcessFunctionTestHarnesses
import org.scalatest.{FlatSpec, Matchers}
import org.swasth.dp.core.job.BaseJobConfig
import org.swasth.dp.core.util.JSONUtil
import org.swasth.dp.coverageeligiblitycheck.functions.CoverageEligibilityProcessFunction
import org.swasth.dp.coverageeligiblitycheck.task.CoverageEligibilityCheckConfig
import org.swasth.fixture.EventFixture

import java.util

class CoverageEligibilityProcessFunctionTest extends FlatSpec with Matchers{

  val config: Config = ConfigFactory.load("coverage-eligibility-check.conf")
  val baseJobConfig: BaseJobConfig = new BaseJobConfig(config,"coverageEligibilityCheck")
  val eligibilityCheckConfig: CoverageEligibilityCheckConfig = new CoverageEligibilityCheckConfig(config)

  //instantiate created user defined function
  val processFunction = new CoverageEligibilityProcessFunction(eligibilityCheckConfig)
  val objMapper = new ObjectMapper
  val gson = new Gson()

  "CoverageEligibilityProcessFunction" should "write data into audit topic as the event does not have context " in {
    // wrap user defined function into a the corresponding operator
    val harness = ProcessFunctionTestHarnesses.forProcessFunction(processFunction)

    //StreamRecord object with the sample data
    val eventMap = gson.fromJson(EventFixture.SAMPLE_EVENT, new util.HashMap[String, AnyRef]().getClass).asInstanceOf[util.Map[String, AnyRef]]
    val testData: StreamRecord[util.Map[String, AnyRef]] = new StreamRecord[util.Map[String, AnyRef]](eventMap)
    harness.processElement(testData)

    harness.getSideOutput(eligibilityCheckConfig.auditOutputTag) should have size 1

    harness.close()
  }

  "CoverageEligibilityProcessFunction" should " have wrong recipient details" in {
    val harness = ProcessFunctionTestHarnesses.forProcessFunction(processFunction)
    val testEvent = JSONUtil.deserialize[util.Map[String, AnyRef]](EventFixture.SAMPLE_INVALID_ACTION_EVENT)
    harness.processElement(testEvent, new java.util.Date().getTime)
  }

}

File: hcx-pipeline-jobs/coverage-eligibility-job/src/test/java/org/swasth/dp/coverageeligibility/functions/CoverageEligibilityProcessFunctionTest.java
package org.swasth.dp.coverageeligibility.functions;

import com.google.gson.Gson;
import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;
import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
import org.apache.flink.streaming.util.OneInputStreamOperatorTestHarness;
import org.apache.flink.streaming.util.ProcessFunctionTestHarnesses;
import org.junit.Test;
import org.swasth.dp.coverageeligiblity.functions.CoverageEligibilityProcessFunction;
import org.swasth.dp.coverageeligiblity.task.CoverageEligibilityConfig;
import org.swasth.fixture.EventFixture;

import java.util.HashMap;
import java.util.Map;

public class CoverageEligibilityProcessFunctionTest {

    Config config  = ConfigFactory.load("coverage-eligibility-test.conf");
    CoverageEligibilityConfig coverageEligibilityConfig = new CoverageEligibilityConfig(config,"CoverageEligibilityTestJob");

    //instantiate created user defined function
    CoverageEligibilityProcessFunction processFunction = new CoverageEligibilityProcessFunction(coverageEligibilityConfig);
    //ObjectMapper objMapper = new ObjectMapper();
    Gson gson = new Gson();


    @Test
    public void testCoverageEligibilityJob() throws Exception {
        // wrap user defined function into the corresponding operator
        OneInputStreamOperatorTestHarness<Map<String, Object>, Map<String, Object>> harness = ProcessFunctionTestHarnesses.forProcessFunction(processFunction);

        //StreamRecord object with the sample data
        Map<String, Object> eventMap = (HashMap<String,Object>) gson.fromJson(EventFixture.SAMPLE_EVENT(),HashMap.class);
        StreamRecord<Map<String, Object>> testData  = new StreamRecord<>(eventMap);
        harness.processElement(testData);
        harness.close();
    }

}